lib/Corres_Cases.thy

```diff
@@ -0,0 +1,319 @@
+
+   R is the predicate to be transformed.
+*)
+private definition corres_case_helper ::
+  "(('a \<Rightarrow> bool) \<times> ('b \<Rightarrow> bool)) \<Rightarrow> (('a \<Rightarrow> bool) \<times> ('b \<Rightarrow> bool)) \<Rightarrow> bool \<Rightarrow> bool" where
+  "corres_case_helper \<equiv> \<lambda>(P, P') (Q, Q') R. (\<forall>s. P s \<longrightarrow> Q s) \<longrightarrow>  (\<forall>s. P' s \<longrightarrow> Q' s) \<longrightarrow> R"
+
+
+(* The following lemmas enable us to lift preconditions of corres_case_helper over conjunction,
+   universal quantifiers, and implication. Note that there are strong versions for forall/implies
+   where both guards are treated uniformly, and weak versions, where forall/implies is dropped
+   in one guard, but not the other.
+*)
+
+private lemma corres_case_helperI:
+  "corres_case_helper (P, P') (P, P') R \<Longrightarrow> R"
+  by (simp add: corres_case_helper_def)
+
+private lemma corres_case_conj_process:
+  "\<lbrakk> corres_case_helper (P, P') (A, A') R; corres_case_helper (P, P') (B, B') S \<rbrakk>
+   \<Longrightarrow> corres_case_helper (P, P') (\<lambda>s. A s \<and> B s, \<lambda>s. A' s \<and> B' s) (R \<and> S)"
+  by (clarsimp simp add: corres_case_helper_def)
+
+private lemma corres_case_all_process:
+  "\<lbrakk> \<And>x. corres_case_helper (P, P') (Q x, Q' x) (R x) \<rbrakk>
+   \<Longrightarrow> corres_case_helper (P, P') (\<lambda>s. \<forall>x. Q x s, \<lambda>s. \<forall>x. Q' x s) (\<forall>x. R x)"
+  by (clarsimp simp: corres_case_helper_def subset_iff)
+
+private lemma corres_case_all_process_weak:
+  "\<lbrakk> \<And>x. corres_case_helper (P, P') (Q x, Q') (R x) \<rbrakk>
+   \<Longrightarrow> corres_case_helper (P, P') (\<lambda>s. \<forall>x. Q x s, Q') (\<forall>x. R x)"
+  by (clarsimp simp: corres_case_helper_def subset_iff)
+
+private lemma corres_case_imp_process:
+  "\<lbrakk> S \<Longrightarrow> corres_case_helper (P, P') (Q, Q') R \<rbrakk>
+   \<Longrightarrow> corres_case_helper (P, P') (\<lambda>s. S \<longrightarrow> Q s, \<lambda>s. S \<longrightarrow> Q' s) (S \<longrightarrow> R)"
+  by (clarsimp simp add: corres_case_helper_def subset_iff)
+
+private lemma corres_case_imp_process_weak:
+  "\<lbrakk> S \<Longrightarrow> corres_case_helper (P, P') (Q, Q') R \<rbrakk>
+   \<Longrightarrow> corres_case_helper (P, P') (\<lambda>s. S \<longrightarrow> Q s, Q') (S \<longrightarrow> R)"
+  by (clarsimp simp add: corres_case_helper_def subset_iff)
+
+private lemmas corres_case_process =
+  corres_case_conj_process corres_case_all_process corres_case_imp_process
+
+private lemmas corres_case_process_weak =
+  corres_case_conj_process corres_case_all_process_weak corres_case_imp_process_weak
+
+(* Turn goals of the form
+
+   (\<forall>y. x = SomeConstr y \<longrightarrow> corres (?P x) P' (SomeConstr y) g) \<and>
+   (\<forall>y. x = OtherConstr y \<longrightarrow> corres (?P x) P' (SomeConstr y) g) \<and>
```

Comment:
is the second one supposed to have OtherConstr y and SomeConstr y?

Reply:
Oh, very well spotted, that should definitely be both `OtherConstr`

Reply:
Ah, it was a bit confusing. Same on the lines below then!